loopout.go

```diff
@@ -1117,121 +1107,23 @@ func (s *loopOutSwap) waitForHtlcSpendConfirmed(globalCtx context.Context,
-					if err != nil {
-						log.Warnf("unable to persist " +
-							"state")
-					}
-
-					return nil, nil
-				}
-
-				// When using taproot HTLCs we're pushing the
-				// preimage before attempting to sweep. This
-				// way the server will know that the swap will
-				// go through and we'll be able to MuSig2
-				// cosign our sweep transaction. In the worst
-				// case if the server is uncooperative for any
-				// reason we can still sweep using scriptpath
-				// spend.
-				err = s.setStatePreimageRevealed(ctx)
-				if err != nil {
-					return nil, err
-				}
-
-				if !paymentComplete {
-					// Push the preimage for as long as the
-					// server is able to settle the swap
-					// invoice. So that we can continue
-					// with the MuSig2 sweep afterwards.
-					s.pushPreimage(ctx)
-				}
-
-				// Now attempt to publish a MuSig2 sweep txn.
-				// Only attempt at most maxMusigSweepRetires
-				// times to still leave time for an emergency
-				// script path sweep.
-				if musigSweepTryCount < maxMusigSweepRetries {
-					success := s.sweepMuSig2(
-						ctx, htlcOutpoint, htlcValue,
-					)
-					if !success {
-						musigSweepTryCount++
-					} else {
-						// Mark that we had a sweep
-						// that was successful. There's
-						// no need for the script spend
-						// now we can just keep pushing
-						// new sweeps to bump the fee.
-						musigSweepSuccess = true
-					}
-				} else if !musigSweepSuccess {
-					// Attempt to script path sweep. If the
-					// sweep fails, we can't do any better
-					// than go on and try again later as
-					// the preimage is already revealed and
-					// the server settled the swap payment.
-					// From the server's point of view the
-					// swap is succeeded at this point so
-					// we are free to retry as long as we
-					// want.
-					err := s.sweep(
-						ctx, htlcOutpoint, htlcValue,
-					)
-					if err != nil {
-						log.Warnf("Failed to publish "+
-							"non-cooperative "+
-							"sweep: %v", err)
-					}
-				}
-
-				// If the result of our spend func was that the
-				// swap has reached a final state, then we
-				// return nil spend details, because there is
-				// no further action required for this swap.
-				if s.state.Type() != loopdb.StateTypePending {
-					return nil, nil
-				}
-			} else {
-				err := s.sweep(ctx, htlcOutpoint, htlcValue)
-				if err != nil {
-					return nil, err
-				}
-
-				// If the result of our spend func was that the
-				// swap has reached a final state, then we
-				// return nil spend details, because there is no
-				// further action required for this swap.
-				if s.state.Type() != loopdb.StateTypePending {
-					return nil, nil
-				}
+			err = s.setStatePreimageRevealed(ctx)
+			if err != nil {
+				return nil, err
+			}
 
-				// If our off chain payment is not yet complete,
-				// we try to push our preimage to the server.
-				if !paymentComplete {
-					s.pushPreimage(ctx)
-				}
+			if !paymentComplete {
```

Comment:
Closed the other thread so we have a clean one to discuss.

- So currently the sweep is added to the sweeper regardless of the cltv. We should at least make sure that it is not timed out yet. (Not checked in the sweeper) [see the now removed `func (s *loopOutSwap) sweepConfTarget()`]
- Then after some time we push the preimage, but the sweeper might already have tried to publish a coop tx, which would fail as the server will not sign the musig2 spend if it hasn't received the preimage yet => so we'd fail back to script sweep. This might not happen in tests, but could very well happen in production. 

What should happen instead:
1. If preimage is not yet revealed we check cltv and if we still can send the sweep to the sweeper.
2. Sweeper should ACK back that the sweep is taken care of.
3. Now we should set the swap's state to preimage revealed (save it) and push the preimage so the server can settle the off-chain payment and will later on sign off any musig2 sweep attempts.
4. Once the server received the preimage and settled the payment (`paymentComplete`) we can ignore the cltv and can just sweep indefinitely as the server won't take the payment back after timeout.
5. Sweeper can just signal back once the sweep is done, no need to send back tx details.